qir/qat/ValidationPass/FunctionValidationPass.cpp

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Logging/ILogger.hpp"
+#include "Profile/Profile.hpp"
+#include "QatTypes/QatTypes.hpp"
+#include "StaticResourceComponent/AllocationAnalysisPass.hpp"
+#include "ValidationPass/FunctionValidationPass.hpp"
+
+#include "Llvm/Llvm.hpp"
+
+#include <functional>
+#include <unordered_map>
+#include <vector>
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    FunctionValidationPass::FunctionValidationPass(ValidationPassConfiguration const& cfg, ILoggerPtr const& logger)
+      : config_{cfg}
+      , logger_{logger}
+    {
+    }
+
+    llvm::PreservedAnalyses FunctionValidationPass::run(llvm::Function& function, llvm::FunctionAnalysisManager& fam)
+    {
+        auto& function_details = fam.getResult<AllocationAnalysisPass>(function);
+
+        if (config_.requiresQubits() && function_details.usage_qubit_counts == 0)
+        {
+            logger_->errorNoQubitsPresent(&function);
+        }
+
+        if (config_.requiresResults() && function_details.usage_result_counts == 0)
+        {
+            logger_->errorNoResultsPresent(&function);
+        }
+
+        return llvm::PreservedAnalyses::all();
+    }
+
+    bool FunctionValidationPass::isRequired()
+    {
+        return true;
+    }
```

Comment:
(personal curiosity)
It is not the first PR of yours where I see this kind of _simple_ implementation for a _static_ member func. Before your PRs I've mostly seen either a more complex implementation (of a static member func) or a non-static member func (typically a virtual func override) with similar simple implementation.
Is the approach that you use a paradigm/template? If yes, then can you please refer me to that template?
What if the instance of this class is manipulated with the base class pointer, and this static member function is called? (I assume the base class also has this member function and it is also static) The bases class' version of this function will be invoked, right? If yes, then what such an approach is intended for, and why is this approach implemented with the _static_ member functions?

Reply:
This is a LLVM function that is mandatory which tells the pass managers whether the pass may be skipped or not. It is used in [PassInstrumentation](https://llvm.org/doxygen/PassInstrumentation_8h_source.html#l00217). I have not read about it in detail and it may be that you can skip adding this function. However, it is unclear what the default behaviour would be without reading the LLVM source in detail.